### PR TITLE
SIRI-96: Provides a rudimentary UI for  BlobHardRefs

### PR DESCRIPTION
**Pull request by andyHa**
The UI provided is a rudimentary approach (mostly copied from storedObjectUploader.html.pasta) as a harsh timeframe was given. This will need some love+work in the near future.

### CODE REVIEW

File: src/main/java/sirius/biz/storage/layer2/Blob.java [hunk 1]
```diff
@@ -160,6 +160,9 @@
      * <p>
      * Also note that if a file is used to provide the new contents of this blob, use
      * {@link #updateContent(String, File)} as this is likely way more efficient.
+     * <p>
+     * Note that this blob object itself will also be update with the aprropriate metadata
```
Comment (mkeckmkeck) [inline src/main/java/sirius/biz/storage/layer2/Blob.java:164]:
```suggestion
     * Note that this blob object itself will also be updated with the appropriate metadata
```

File: src/main/java/sirius/biz/storage/layer2/BlobController.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+     * as permanent.
+     *
+     * @param ctx       the request to handle
+     * @param out       the response to the AJAX call
+     * @param spaceName the {@link BlobStorageSpace} to store the object in
+     * @param upload    the content of the upload
+     */
+    @Routed(value = "/dasd/upload-file/:1", preDispatchable = true, jsonCall = true)
+    @LoginRequired
+    public void uploaduploadFile(final WebContext ctx,
```
Comment (mkeckmkeck) [inline src/main/java/sirius/biz/storage/layer2/BlobController.java:53]:
```suggestion
    public void uploadFile(final WebContext ctx,
```

File: src/main/java/sirius/biz/storage/layer2/BlobHardRefProperty.java [hunk 3]
```diff
@@ -151,13 +151,14 @@ protected void onBeforeSaveChecks(Object entity) {
...
-               .getStorageSpace()
-               .deleteReferencedBlobs(((BaseEntity<?>) entity).getUniqueName(), getName(), ref.getKey());
+        if (ref.changed) {
+            BlobStorageSpace storageSpace = ref.getBlob().getStorageSpace();
+            String uniqueName = ((BaseEntity<?>) entity).getUniqueName();
+
+            if (ref.isFilled()) {
+                storageSpace.markAsUsed(uniqueName, getName(), ref.getKey());
+            }
+            storageSpace.deleteReferencedBlobs(uniqueName, getName(), ref.getKey());
```
Comment (sabieber) [inline src/main/java/sirius/biz/storage/layer2/BlobHardRefProperty.java:161]: This also changes the logic not only readability??

File: src/main/java/sirius/biz/storage/layer2/mongo/MongoBlobStorageSpace.java [hunk 4]
```diff
@@ -341,7 +341,14 @@ protected MongoDirectory fetchBlobParent(MongoBlob blob) {
                                      .executeFor(MongoBlob.class)
                                      .getModifiedCount();
             if (numUpdated == 1) {
-                return Optional.ofNullable(blob.getPhysicalObjectKey());
+                String previousPhysicalObjectKey = blob.getPhysicalObjectKey();
```
Comment (sabieber) [inline src/main/java/sirius/biz/storage/layer2/mongo/MongoBlobStorageSpace.java:344]: Das Kommentar von oben übernehmen?

File: src/main/java/sirius/biz/storage/layer2/BlobController.java [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+ * {@link sirius.biz.storage.layer3.VFSController}.
+ */
+@Register(framework = StorageUtils.FRAMEWORK_STORAGE, classes = Controller.class)
+public class BlobController extends BizController {
+
+    @Part
+    private BlobStorage blobStorage;
+
+    /**
+     * Uploads a file into a temporary Blob, to be later persisted via a {@link BlobHardRef}.
```
Comment (sabieber) [inline src/main/java/sirius/biz/storage/layer2/BlobController.java:40]: link blob?

File: src/main/java/sirius/biz/storage/layer2/BlobController.java [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+    @LoginRequired
+    public void uploaduploadFile(final WebContext ctx,
+                                 JSONStructuredOutput out,
+                                 String spaceName,
+                                 InputStreamHandler upload) {
+        Blob blob = blobStorage.getSpace(spaceName).createTemporaryBlob();
+        try {
+            try {
+                ctx.markAsLongCall();
+                //TODO remove legacy qqfile once library is updated...
```
Comment (sabieber) [inline src/main/java/sirius/biz/storage/layer2/BlobController.java:61]: TODO ohne Ticket referenz?

File: src/main/java/sirius/biz/storage/layer2/BlobController.java [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+                String name = ctx.get("filename").asString(ctx.get("qqfile").asString());
+                blob.updateContent(name, upload, Long.parseLong(ctx.getHeader(HttpHeaderNames.CONTENT_LENGTH)));
+
+                out.property("fileId", blob.getBlobKey());
+
+                // TODO remove once the blobHardRefField has been refactored
```
Comment (sabieber) [inline src/main/java/sirius/biz/storage/layer2/BlobController.java:67]: TODO ohne Ticket Referenz?

File: src/main/java/sirius/biz/storage/layer2/mongo/MongoBlobStorageSpace.java (hunk 4 above)
Comment (sabieber) [reply src/main/java/sirius/biz/storage/layer2/mongo/MongoBlobStorageSpace.java:344]: Oder kann man das vllt in ne Methode auslagern?

**Pull request merged**